test/jdk/javax/net/ssl/SSLSession/ReadOnlyEngine.java

```diff
@@ -0,0 +1,214 @@
+    }
+
+    HandshakeStatus receive(SSLEngine engine, ByteBuffer src, ByteBuffer dst)
+        throws SSLException {
+        SSLEngineResult status = engine.unwrap(src, dst);
+        dst.flip();
+        return status.getHandshakeStatus();
+    }
+
+    ReadOnlyEngine(SSLContext sslc, String ciphersuite) throws Exception {
+        System.err.println("==== Test Protocol: " + sslc.getProtocol() +
+            ", CipherSuite: " + ciphersuite);
+        KeyStore ks = KeyStore.getInstance("PKCS12");
+        KeyStore ts = KeyStore.getInstance("PKCS12");
+
+        ks.load(new FileInputStream(keyFilename), passwd);
+        ts.load(new FileInputStream(trustFilename), passwd);
+
+        KeyManagerFactory kmf = KeyManagerFactory.getInstance("SunX509");
+        kmf.init(ks, passwd);
+        TrustManagerFactory tmf = TrustManagerFactory.getInstance("SunX509");
+        tmf.init(ts);
+
+        sslc.init(kmf.getKeyManagers(), tmf.getTrustManagers(), null);
+        client = sslc.createSSLEngine("client", 1);
+        client.setUseClientMode(true);
+        server = sslc.createSSLEngine("server", 2);
+        if (ciphersuite != null) {
+            server.setEnabledCipherSuites(new String[] { ciphersuite });
+        }
+        server.setUseClientMode(false);
+
+        SSLSession session = server.getSession();
+        int maxData = session.getPacketBufferSize();
+
+        ByteBuffer in = ByteBuffer.allocate(maxData);
+        ByteBuffer out = ByteBuffer.allocate(maxData);
+
+        HandshakeStatus statusClient, statusServer;
+        client.beginHandshake();
+        server.beginHandshake();
+
+        // Do TLS handshake
+        do {
+            statusClient = doHandshake(client, out, in);
+            statusServer = doHandshake(server, in, out);
+        } while (statusClient != HandshakeStatus.NOT_HANDSHAKING ||
+            statusServer != HandshakeStatus.NOT_HANDSHAKING);
+
+        // Read NST
+        in.clear();
+        receive(client, out, in);
+
+        System.out.println("done");
+
+        // Send bytes from the client and make sure the server receives the same
+        in.clear();
+        out.clear();
+        String testString = "ASDF";
+        in.put(testString.getBytes()).flip();
+        String testResult;
+        System.out.println("1: Client send: " + testString);
+        send(client, in.asReadOnlyBuffer(), out);
+        in.clear();
+        receive(server, out.asReadOnlyBuffer(), in);
+        testResult = StandardCharsets.UTF_8.decode(in.duplicate()).toString();
+        System.out.println("1: Server receive: " + testResult);
+        if (!testString.equalsIgnoreCase(testResult)) {
+            throw new Exception("unequal");
+        }
+
+        // Send bytes from the server and make sure the client receives the same
+        out.clear();
+        in.clear();
+        System.out.println("2: Server send: " + testString);
+        in.put(testString.getBytes()).flip();
+        send(server, in, out);
```

Comment:
Did you want to try asReadOnlyBuffer() here also?

Reply:
Yeah, they should have been.  I must have undid it to test something